File: src/app/weapons/projectile/projectile.lua
require("lualib_bundle");
local ____exports = {}
local ____projectile_2Dtarget = require("app.weapons.projectile.projectile-target")
local ProjectileMoverLinear = ____projectile_2Dtarget.ProjectileMoverLinear
local ____projectile_2Dsfx = require("app.weapons.projectile.projectile-sfx")
local ProjectileSFX = ____projectile_2Dsfx.ProjectileSFX
local DEFAULT_FILTER
DEFAULT_FILTER = function(projectile)
    return Filter(
        function()
            local unit = GetFilterUnit()
            return GetWidgetLife(unit) > 0.405 and not IsUnitAlly(
                unit,
                GetOwningPlayer(projectile.source)
            ) and IsUnitType(unit, UNIT_TYPE_STRUCTURE) == false and IsUnitType(unit, UNIT_TYPE_MAGIC_IMMUNE) == false and IsUnitType(unit, UNIT_TYPE_MECHANICAL) == false
        end
    )
end
____exports.Projectile = {}
local Projectile = ____exports.Projectile
Projectile.name = "Projectile"
Projectile.__index = Projectile
Projectile.prototype = {}
Projectile.prototype.__index = Projectile.prototype
Projectile.prototype.constructor = Projectile
function Projectile.new(...)
    local self = setmetatable({}, Projectile.prototype)
    self:____constructor(...)
    return self
end
function Projectile.prototype.____constructor(self, source, startPosition, target, projectileMover)
    self.collisionRadius = 30
    self.velocity = 10
    self.doDestroy = false
    self.position = startPosition
    self.target = target
    self.sfx = {}
    self.mover = projectileMover or ProjectileMoverLinear.new()
    self.source = source
    self.filter = DEFAULT_FILTER(self)
end
function Projectile.prototype.addEffect(self, sfx, offset, facing, scale)
    local _sfx = ProjectileSFX.new(sfx, self.position, offset, facing)
    _sfx:setScale(scale)
    __TS__ArrayPush(self.sfx, _sfx)
    return _sfx:getEffect()
end
function Projectile.prototype.doesCollide(self)
    return true
end
function Projectile.prototype.getPosition(self)
    return self.position
end
function Projectile.prototype.getCollisionRadius(self)
    return self.collisionRadius
end
function Projectile.prototype.setCollisionRadius(self, radius)
    self.collisionRadius = radius
    return self
end
function Projectile.prototype.willDestroy(self)
    return self.doDestroy
end
function Projectile.prototype.setDestroy(self, val)
    self.doDestroy = val
    return self
end
function Projectile.prototype.overrideFilter(self, newFunc)
    self.filter = Filter(newFunc)
    return self
end
function Projectile.prototype.onCollide(self, callback)
    self.onCollideCallback = callback
    return self
end
function Projectile.prototype.onDeath(self, callback)
    self.onDeathCallback = callback
    return self
end
function Projectile.prototype.collide(self, weaponModule, withUnit)
    if self.onCollideCallback then
        self:onCollideCallback(weaponModule, self, withUnit)
    end
end
function Projectile.prototype.getTarget(self)
    return self.target
end
function Projectile.prototype.update(self, weaponModule, deltaTime)
    local velocityToApply = self.mover:move(
        self.position,
        self:getTarget():getTargetVector(),
        self.velocity,
        deltaTime
    )
    local newPosition = self.position:add(velocityToApply)
    self.position = newPosition
    __TS__ArrayForEach(
        self.sfx,
        function(____, sfx) return sfx:updatePosition(self.position) end
    )
    if self:reachedEnd(weaponModule, velocityToApply) then
        self.doDestroy = true
    end
    return velocityToApply
end
function Projectile.prototype.setVelocity(self, velocity)
    self.velocity = velocity
    return self
end
function Projectile.prototype.reachedEnd(self, weaponModule, targetVector)
    MoveLocation(weaponModule.game.TEMP_LOCATION, self.position.x, self.position.y)
    local z = GetLocationZ(weaponModule.game.TEMP_LOCATION)
    return (self.position.z <= z)
end
function Projectile.prototype.destroy(self)
    local ____ = self.onDeathCallback and self:onDeathCallback(self)
    __TS__ArrayForEach(
        self.sfx,
        function(____, sfx) return sfx:destroy() end
    )
    self.sfx = {}
    return true
end
return ____exports

File: src/app/weapons/projectile/projectile.ts
import { ProjectileTarget, ProjectileMover, ProjectileMoverLinear } from "./projectile-target";
import { Vector3 } from "../../types/vector3";
import { ProjectileSFX } from "./projectile-sfx";
import { getZFromXY } from "lib/utils";
import { UNIT_IS_FLY } from "resources/ability-ids";

const DEFAULT_FILTER = (projectile: Projectile) => {
    return Filter(() => {
        let unit = GetFilterUnit(); 
        
        return GetWidgetLife(unit) > 0.405 && 
            !IsUnitAlly(unit, GetOwningPlayer(projectile.source)) &&
            IsUnitType(unit, UNIT_TYPE_MAGIC_IMMUNE) == false && 
            BlzIsUnitInvulnerable(unit) === false &&
            GetUnitAbilityLevel(unit, UNIT_IS_FLY) == 0;
    });
};

export class Projectile {
    private id = 0;
    private position: Vector3;

    private target: ProjectileTarget;
    private mover: ProjectileMover;
    
    public source: unit;

    private collisionRadius: number = 30;

    private sfx: Array<ProjectileSFX>;

    private doDestroy = false;
    public dead = false;

    public filter: filterfunc;


    private stopAtGroundCollision = true;
    private stopAtGoal = false;

    // Callbacks
    private onCollideCallback: ((projectile: Projectile, who: unit) => boolean | void) | undefined;
    private onDeathCallback: ((projectile: Projectile) => boolean | void) | undefined;

    constructor(source: unit, startPosition: Vector3, target: ProjectileTarget, projectileMover?: ProjectileMover) {
        this.position = startPosition;
        this.target = target;
        this.sfx = [];

        this.mover = projectileMover || new ProjectileMoverLinear(startPosition, target.getTargetVector());

        this.source = source;
        this.filter = DEFAULT_FILTER(this);
    }

    setId(to: number) {
        this.id = to;
    }

    getId() {
        return this.id;
    }

    /**
     * Adds an effect to the projectile
     * @param sfx 
     * @param offset 
     * @param facing 
     */
    addEffect(sfx: string, offset: Vector3, facing: Vector3, scale: number): effect {
        let _sfx = new ProjectileSFX(sfx, this.position, offset, facing);
        _sfx.setScale(scale);
        this.sfx.push(_sfx);
        return _sfx.getEffect();
    }

    public getSfx() {
        return this.sfx
    }

    /**
     * Does the projectile collide
     */
    doesCollide(): boolean {
        return true;
    }
    /**
     * Returns the projctile's current position
     */
    getPosition(): Vector3 {
        return this.position;
    }
    
    /**
     * Gets collision radius of the projectile
     */
    getCollisionRadius(): number {
        return this.collisionRadius;
    }

    setCollisionRadius(radius: number): Projectile {
        this.collisionRadius = radius;
        return this;
    }

    willDestroy(): boolean {
        return this.doDestroy;
    }

    setDestroy(val: boolean): Projectile {
        this.doDestroy = val;
        return this;
    }

    public overrideFilter(newFunc: (projectile: Projectile) => boolean): Projectile {
        this.filter = Filter(<any>newFunc);
        return this;
    }

    public onCollide(callback: (projectile: Projectile, who: unit) => boolean | void): Projectile {
        this.onCollideCallback = callback;
        return this;
    }

    public onDeath(callback: (projectile: Projectile) => boolean | void): Projectile {
        this.onDeathCallback = callback;
        return this;
    }

    public collide(withUnit: unit): void {
        if (this.onCollideCallback) {
            this.onCollideCallback(this, withUnit);
        }
    }
    
    public getTarget(): ProjectileTarget {
        return this.target;
    }
    /**
     * Updates the projectile's location and returns distance moved
     * @param deltaTime 
     */
    update(deltaTime: number): Vector3 {
        { // DO
            let velocityToApply = this.mover.move(this.getTarget().getTargetVector(), deltaTime);
            let newPosition = this.position.add(velocityToApply);
            this.position = newPosition;

            // Now update attached 
            this.sfx.forEach(sfx => sfx.updatePosition(this.position, velocityToApply));
            
            if (this.reachedEnd(velocityToApply)) this.doDestroy = true;
            // Return distance travelled
            return velocityToApply;
        } // END
    }

    public setVelocity(velocity: number): Projectile {
        this.mover.setVelocity(velocity);
        return this;
    }

    private reachedEnd(targetVector: Vector3): boolean {
        if (this.stopAtGoal) {
            return this.mover.reachedGoal();
        }
        if (this.stopAtGroundCollision) {
            let z = getZFromXY(this.position.x, this.position.y);
            // let location = GLOBAL_LOCATION
            return (this.position.z <= z);
        }
        return true;
    }

    public doStopAtGoal(state: boolean) {
        this.stopAtGoal = state;
        return this;
    }
    public doStopAtGroundCollision(state: boolean) {
        this.stopAtGroundCollision = state;
        return this;
    }

    public destroy(): boolean {
        if (this.onDeathCallback) {
            let result = this.onDeathCallback(this);
            if (result == false) {
                this.dead = result 
            }
            else {
                this.dead = true;
            }
        }
        else {
            this.dead = true;
        }
        
        if (this.dead) {
            this.sfx.forEach(sfx => sfx.destroy());
            this.sfx = [];
            if (this.filter) {
                DestroyFilter(this.filter);
            }
        }
        return this.dead;
    }

    public getDoodadChecker() {
        return this.mover.getDoodadChecker();
    }
}
